File: api/python-ai/generate_test_interactions.py
"""
Generate test interaction data for current logged-in user
This creates sample interactions so you can see ML recommendations immediately
"""

import os
import sys
from datetime import datetime, timedelta
import random
from dotenv import load_dotenv
from pymongo import MongoClient

# Load environment variables
load_dotenv()

# MongoDB connection
MONGO_URI = os.getenv('MONGO_URI')
if not MONGO_URI:
    print("Error: MONGO_URI not found in environment variables")
    sys.exit(1)

client = MongoClient(MONGO_URI)
db = client['FocusDesk']

def generate_interactions_for_user(user_id, num_interactions=50):
    """Generate test interactions for a specific user"""
    
    # Get available packages
    packages = list(db.packages.find({'isActive': True}).limit(20))
    if not packages:
        print("No packages found in database!")
        return
    
    print(f"Found {len(packages)} packages")
    print(f"Generating {num_interactions} interactions for user: {user_id}")
    
    interactions = []
    interaction_types = ['view', 'click', 'view', 'click', 'view']  # More views and clicks
    
    # Generate interactions over the past 30 days
    base_time = datetime.now()
    
    for i in range(num_interactions):
        # Random package
        package = random.choice(packages)
        
        # Random interaction type
        interaction_type = random.choice(interaction_types)
        
        # Random timestamp in the past 30 days
        days_ago = random.randint(0, 30)
        hours_ago = random.randint(0, 23)
        timestamp = base_time - timedelta(days=days_ago, hours=hours_ago)
        
        interaction = {
            'userId': user_id,
            'packageId': str(package['_id']),
            'interactionType': interaction_type,
            'timestamp': timestamp,
            'metadata': {
                'source': 'test_data_generation',
                'packageTitle': package.get('title', 'Unknown')
            }
        }
        
        interactions.append(interaction)
    
    # Insert interactions
    if interactions:
        result = db.interactions.insert_many(interactions)
        print(f"✅ Created {len(result.inserted_ids)} interactions")
        
        # Show summary
        print("\nInteraction Summary:")
        type_counts = {}
        for interaction in interactions:
            t = interaction['interactionType']
            type_counts[t] = type_counts.get(t, 0) + 1
        
        for itype, count in type_counts.items():
            print(f"  - {itype}: {count}")
        
        # Show some package titles
        print("\nPackages interacted with:")
        unique_packages = set(i['metadata']['packageTitle'] for i in interactions[:10])
        for title in list(unique_packages)[:5]:
            print(f"  - {title}")
    
    client.close()
    print("\n✅ Done! Now retrain the model with: python train_model.py")

if __name__ == '__main__':
    # Your actual user ID from the console logs
    USER_ID = '680a9116b88e07cba6d4572b'
    
    # Allow custom user ID from command line
    if len(sys.argv) > 1:
        USER_ID = sys.argv[1]
    
    # Allow custom number of interactions
    num_interactions = 50
    if len(sys.argv) > 2:
        num_interactions = int(sys.argv[2])
    
    print(f"Generating test interactions for user: {USER_ID}")
    generate_interactions_for_user(USER_ID, num_interactions)

File: api/scripts/generate-test-interactions.js
/**
 * Generate test interactions for a user to populate ML recommendations
 * Run with: node generate-test-interactions.js <userId> <numInteractions>
 */

import mongoose from 'mongoose';
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load environment variables
dotenv.config({ path: path.join(__dirname, '../.env') });

// Define interaction schema
const interactionSchema = new mongoose.Schema({
  userId: { type: String, required: true },
  packageId: { type: String, required: true },
  interactionType: { type: String, enum: ['view', 'click', 'search', 'message', 'rating', 'booking', 'start_booking'], required: true },
  timestamp: { type: Date, default: Date.now },
  metadata: mongoose.Schema.Types.Mixed
});

const Interaction = mongoose.model('Interaction', interactionSchema);

// Package schema (simplified)
const packageSchema = new mongoose.Schema({}, { strict: false });
const Package = mongoose.model('Package', packageSchema, 'packages');

async function generateInteractions(userId, numInteractions = 50) {
  try {
    // Connect to MongoDB
    console.log('Connecting to MongoDB...');
    await mongoose.connect(process.env.MONGO_URI);
    console.log('✅ Connected to MongoDB');

    // Get active packages
    const packages = await Package.find({ isActive: true }).limit(20);
    if (packages.length === 0) {
      console.log('❌ No packages found in database!');
      return;
    }

    console.log(`Found ${packages.length} packages`);
    console.log(`Generating ${numInteractions} interactions for user: ${userId}\n`);

    const interactions = [];
    const interactionTypes = ['view', 'view', 'view', 'click', 'click', 'view'];
    
    // Generate interactions over the past 30 days
    const now = new Date();
    
    for (let i = 0; i < numInteractions; i++) {
      // Random package
      const randomPkg = packages[Math.floor(Math.random() * packages.length)];
      
      // Random interaction type
      const interactionType = interactionTypes[Math.floor(Math.random() * interactionTypes.length)];
      
      // Random timestamp in the past 30 days
      const daysAgo = Math.floor(Math.random() * 30);
      const hoursAgo = Math.floor(Math.random() * 24);
      const timestamp = new Date(now.getTime() - (daysAgo * 24 * 60 * 60 * 1000) - (hoursAgo * 60 * 60 * 1000));
      
      interactions.push({
        userId,
        packageId: randomPkg._id.toString(),
        interactionType,
        timestamp,
        metadata: {
          source: 'test_data_generation',
          packageTitle: randomPkg.title || 'Unknown'
        }
      });
    }

    // Insert interactions
    const result = await Interaction.insertMany(interactions);
    console.log(`✅ Created ${result.length} interactions\n`);

    // Show summary
    console.log('Interaction Summary:');
    const typeCounts = {};
    interactions.forEach(interaction => {
      const t = interaction.interactionType;
      typeCounts[t] = (typeCounts[t] || 0) + 1;
    });

    Object.entries(typeCounts).forEach(([type, count]) => {
      console.log(`  - ${type}: ${count}`);
    });

    // Show some package titles
    console.log('\nPackages interacted with:');
    const uniquePackages = [...new Set(interactions.slice(0, 10).map(i => i.metadata.packageTitle))];
    uniquePackages.slice(0, 5).forEach(title => {
      console.log(`  - ${title}`);
    });

    console.log('\n✅ Done! Now retrain the model with:');
    console.log('   cd python-ai');
    console.log('   python train_model.py');

  } catch (error) {
    console.error('❌ Error:', error.message);
  } finally {
    await mongoose.connection.close();
    console.log('\n Database connection closed');
  }
}

// Get command line arguments
const userId = process.argv[2] || '680a9116b88e07cba6d4572b';
const numInteractions = parseInt(process.argv[3]) || 50;

console.log(`Generating test interactions for user: ${userId}`);
generateInteractions(userId, numInteractions);
